File: 김수미/boj/G5_1956.py
'''
 * Author    : Kim Su Mi
 * Date      : 2024.09.28(SAT)
 * Runtime   : 764 ms
 * Memory    : 111912 KB
 * Algorithm : Floyd-Warshall
'''

import sys
v, e = map(int, sys.stdin.readline().split())

# 방향성 있는 그래프에서 최단거리 구하기
graph = [[int(1e9)] * (v+1) for _ in range(v+1)]

# 도로의 길이를 값으로 넣기
for _ in range(e):
    a, b, c = map(int, sys.stdin.readline().split())
    graph[a][b] = c

# 플로이드 와샬 : 모든 노드에서 다른 모든 노드까지의 최단경로 계산
for i in range(1, v+1):
    for a in range(1, v+1):
        for b in range(1, v+1):
            graph[a][b] = min(graph[a][b], graph[a][i] + graph[i][b])

# 자기 자신마을에서 자기 자신마을으로 돌아오는 거리 중 최소값 찾기
result = int(1e9)
for i in range(1, v+1):
    result = min(result, graph[i][i])

if result == int(1e9):
    print(-1)
else:
    print(result)

File: 김수미/boj/G5_1956.java
package 김수미.boj;

/**
 * Author    : Kim Su Mi
 * Date      : 2024.09.28(SAT)
 * Runtime   : 476 ms
 * Memory    : 58456 KB
 * Algorithm : Floyd-Warshall
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class G5_1956 {

	static final int INF = 987654321;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		StringTokenizer st = new StringTokenizer(br.readLine());
		int v = Integer.parseInt(st.nextToken());
		int e = Integer.parseInt(st.nextToken());

		int[][] graph = new int[v+1][v+1];
		for (int[] row : graph) Arrays.fill(row, INF);

		while (e-->0) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			graph[a][b] = c;
		}

		for(int i=1; i<v+1; i++){
			for(int a=1; a<v+1; a++){
				for(int b=1; b<v+1; b++){
					graph[a][b] = Math.min(graph[a][b], graph[a][i] + graph[i][b]);
				}
			}
		}

		int result = INF;
		for(int i=1; i<v+1; i++){
			result = Math.min(result, graph[i][i]);
		}

		System.out.println(result == INF ? -1 : result);

	}
}
